### PR TITLE
Port Applicative to Dotty

### PR DESCRIPTION
**Pull request by andrzejressel**
Requires Scala [3.1.2-RC1](https://github.com/lampepfl/dotty/releases/tag/3.1.2-RC1)

### CODE REVIEW

**Comment by andrzejressel** [comment]
I was some tuple fixes in 3.1.1 - maybe that will help

**Comment by andrzejressel** [comment]
I've commented out Tuple examples and added comment, so it can be merged as is, without any waiting.

File: core/src/main/scala-3/cats/derived/DerivedApplicative.scala [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    def ap[A, B](ff: T)(fa: T) = T.combine(ff, fa)
+
+  given [F[_], G[_]](using F: Applicative[F], G: Applicative[G]): DerivedApplicative[[x] =>> F[G[x]]] =
+    F.compose(G)
+
+  given [F[_]](using inst: => K1.ProductInstances[Or, F]): DerivedApplicative[F] =
+    given K1.ProductInstances[Applicative, F] = inst.unify
+    new Product[Applicative, F] {}
+
+  trait Product[T[x[_]] <: Applicative[x], F[_]](using inst: K1.ProductInstances[T, F]) extends Applicative[F]:
```
Comment (joroKr21) [inline core/src/main/scala-3/cats/derived/DerivedApplicative.scala:38]: Actually the idea is that you can reuse your `Apply` implementation so you should add `extends DerivedApply.Product[T, F]` here

Comment (andrzejressel) [reply core/src/main/scala-3/cats/derived/DerivedApplicative.scala:38]: Yeah, I've realized that after I finished Apply :D

**Comment by joroKr21** [comment]
It's not a big deal about the tests - it's just satisfying if it works because it makes it easier to add another test. But if it doesn't work, oh well we lived with it in Scala 2.

**Review by joroKr21 — APPROVED** [review]
LGTM 👍

**Pull request merged**